5/src/main/java/com/whoisacat/edu/book/catalogue/dao/BookDaoJdbc.java

```diff
@@ -0,0 +1,150 @@
+package com.whoisacat.edu.book.catalogue.dao;
+
+import com.whoisacat.edu.book.catalogue.domain.Author;
+import com.whoisacat.edu.book.catalogue.domain.Book;
+import com.whoisacat.edu.book.catalogue.domain.Genre;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcOperations;
+import org.springframework.stereotype.Repository;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+@SuppressWarnings({"SqlNoDataSourceInspection", "ConstantConditions", "SqlDialectInspection"})
+@Repository
+public class BookDaoJdbc implements BookDao{
+
+    private final NamedParameterJdbcOperations jdbc;
+
+    public BookDaoJdbc(NamedParameterJdbcOperations namedParameterJdbcOperations)
+    {
+        this.jdbc = namedParameterJdbcOperations;
+    }
+
+    @Override
+    public int countAll() {
+        return jdbc.getJdbcOperations().queryForObject("select count(*) from book",Integer.class);
+    }
+
+    @Override
+    public int insert(Book book) {
+        Map<String, Object> params = Map.of("id",getNewId(),
+                "name", book.getName(),
+                "author_id",book.getAuthor().getId(),
+                "genre_id",book.getGenre().getId());
+        return jdbc.update("insert into book (id, \"name\",author_id,genre_id) " +
+                "values (:id, :name,:author_id,:genre_id)", params);
+    }
+
+    private long getNewId(){
+        return jdbc.getJdbcOperations().queryForObject("select max(id) + 1 from book",Integer.class);
+    }
+
+    @Override
+    public Book getById(long id) {
+        MapSqlParameterSource params = new MapSqlParameterSource("id", id);
+        Book book = null;
+        try{
+
+            book = jdbc.queryForObject(
+                    "select bk.id as bkid,bk.\"name\" as bkname, " +
+                            "au.id as auid,au.\"name\" as auname, " +
+                            "gr.id as grid,gr.\"name\" as grname " +
+                            "from book bk " +
+                            "left join author au on bk.author_id = au.id " +
+                            "left join genre gr on bk.genre_id = gr.id " +
+                            "where bk.id = :id;", params, new BookMapper());
+        }finally{
+            return book;
+        }
```

Comment:
>Половина спринга на этом построена. 
Ого. Не встречал( Можете привести пример?

>Блок finally нужен чтобы гарантированно возвращать значение. Для того он тут и используется.

Он нужен не для этого
https://docs.oracle.com/javase/tutorial/essential/exceptions/summary.html
>The finally block identifies a block of code that is guaranteed to execute, and is the right place to close files, recover resources, and otherwise clean up after the code enclosed in the try block.


Следующий код фактически аналогичен, тому что в выполненном задании. Неужели ничего глаз не режет?
```
    private static String anyAction() {
        String s = "Нуничегосебе!";
        try {
            throw new OutOfMemoryError();
        } finally {
            return s;
        }
    }

    public static void main(String[] args) throws Exception {
        System.out.println(anyAction());
    }
```

Мы подавляем исключения. Все. Тем самым нивелируя весь тот профит ради которого они и были созданы
https://docs.oracle.com/javase/tutorial/essential/exceptions/advantages.html

Reply:
поправил. теперь мне не нравится что я написал. вариант ноль книг теперь не рассматривается. самому не нравится, а что делать не знаю.